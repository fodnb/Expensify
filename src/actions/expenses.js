import uuid from "uuid";
import database from "../../src/firebase/firebase";
//components call action generator
//action generator returns object
//component calls action generator dispatches function
// function runs (has the ability to dispatch other and do whatever it wants)


export const addExpense = (expense) => ({
    type: "ADD_EXPENSE",
    expense
})

// REMOVE EXPENSE
export const removeExpense = ({id} = {}) => ({
    type: "REMOVE_EXPENSE",
    id 
});

export const editExpense = (id, updates) => ({
    type: "EDIT_EXPENSE",
    id,
    updates
})

export const startAddExpense = (expenseData = {}) => {
    return (dispatch)=>{
        const {
            description = "",
            note = '',
            createdAt = 0,
            amount = 0
        } = expenseData;

        const expense = { description, note, createdAt, amount};

       return database.ref('expenses').push(expense)
            .then((snapshot)=>{
                dispatch(addExpense({
                    id: snapshot.key,
                    ...expense
                }))
            });


    }
}