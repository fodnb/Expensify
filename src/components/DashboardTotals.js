// import selectExpensesTotals from '../selectors/expenses-total.js';


// const expenses = [{
//     id: "1",
//     description: "gum",
//     note: '',
//     amount: 195
// },
// {
//     id: "2",
//     description: "rent",
//     note: '',
//     amount: 170000.00
// },
// {
//     id: "3",
//     description: "credit card",
//     note: '',
//     amount: 4000
// }    
// ]

export default(expenses) =>{

const getExpenseTotal = (expenses) =>{
    expenseTotal = expenses.map((expense)=> {
        return totalsArray.push(expense.amount);
    })
    total = totalsArray.reduce((prev, cur)=> {
        return total = prev + cur;
    })
    
    return total
}

let totalsArray = [], total, expenseTotal;

{expenses[0].amount ? getExpenseTotal(expenses) : total = 0}

return total;

}