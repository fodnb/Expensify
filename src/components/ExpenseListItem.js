import React from 'react';
import {Link} from 'react-router-dom';

const ExpenseListItem = ({ id, description, createdAt, amount}) => (
    <div>
        <h3><Link to={`/edit/${id}`}>{description}</Link></h3>
        <p>{createdAt}</p>
        <p>{amount}</p>
        
    </div>
);

export default ExpenseListItem;  