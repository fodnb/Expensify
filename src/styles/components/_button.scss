.button {
    background: $blue;
    color: white;
    padding: $s-size;
    border: none;
    margin: 0;
    font: {
        size: $font-size-large;
        weight: 300;
    }
}

.button--link {
    background: none;
}

.secondary {
    background: #888;
}