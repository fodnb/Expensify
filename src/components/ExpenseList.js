import React from "react";
import { connect } from 'react-redux';
import ExpenseListItem  from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";

export const ExpenseList = (props) => (

    <div>
        {props.expenses.length === 0 ? (<p> No Expenses </p>): (
            props.expenses.map((expense)=>(
                <ExpenseListItem key={expense.id} {...expense}/>
            ))   
        )}
        <h1>expense list</h1>
        
    </div>
);

const mapStoreToProps = (state)=>{
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
}

export default connect(mapStoreToProps)(ExpenseList);
