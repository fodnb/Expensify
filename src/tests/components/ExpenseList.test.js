import React from "react";
import {shallow} from 'enzyme';
import { ExpenseList } from '../../components/ExpenseList';
import expenses from "../fixtures/expenses";


test('should render expenselist with expense', ()=> {

    const wrapper = shallow(<ExpenseList expenses={expenses}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should render no expense p element with no expenses',()=>{
    const wrapper = shallow(<ExpenseList expenses={[]} />);

    expect(wrapper).toMatchSnapshot();

})
