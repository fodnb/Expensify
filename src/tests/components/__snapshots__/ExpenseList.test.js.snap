// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expenselist with expense 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ExpenseList
    expenses={
      Array [
        Object {
          "amount": 195,
          "createdAt": 0,
          "description": "gum",
          "id": "1",
          "note": "",
        },
        Object {
          "amount": 170000,
          "createdAt": -345600000,
          "description": "rent",
          "id": "2",
          "note": "",
        },
        Object {
          "amount": 4000,
          "createdAt": 345600000,
          "description": "credit card",
          "id": "3",
          "note": "",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        Array [
          <ExpenseListItem
            amount={195}
            createdAt={0}
            description="gum"
            id="1"
            note=""
          />,
          <ExpenseListItem
            amount={170000}
            createdAt={-345600000}
            description="rent"
            id="2"
            note=""
          />,
          <ExpenseListItem
            amount={4000}
            createdAt={345600000}
            description="credit card"
            id="3"
            note=""
          />,
        ],
        <h1>
          expense list
        </h1>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "function",
        "props": Object {
          "amount": 195,
          "createdAt": 0,
          "description": "gum",
          "id": "1",
          "note": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "2",
        "nodeType": "function",
        "props": Object {
          "amount": 170000,
          "createdAt": -345600000,
          "description": "rent",
          "id": "2",
          "note": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "3",
        "nodeType": "function",
        "props": Object {
          "amount": 4000,
          "createdAt": 345600000,
          "description": "credit card",
          "id": "3",
          "note": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "expense list",
        },
        "ref": null,
        "rendered": "expense list",
        "type": "h1",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          Array [
            <ExpenseListItem
              amount={195}
              createdAt={0}
              description="gum"
              id="1"
              note=""
            />,
            <ExpenseListItem
              amount={170000}
              createdAt={-345600000}
              description="rent"
              id="2"
              note=""
            />,
            <ExpenseListItem
              amount={4000}
              createdAt={345600000}
              description="credit card"
              id="3"
              note=""
            />,
          ],
          <h1>
            expense list
          </h1>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "function",
          "props": Object {
            "amount": 195,
            "createdAt": 0,
            "description": "gum",
            "id": "1",
            "note": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "2",
          "nodeType": "function",
          "props": Object {
            "amount": 170000,
            "createdAt": -345600000,
            "description": "rent",
            "id": "2",
            "note": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "3",
          "nodeType": "function",
          "props": Object {
            "amount": 4000,
            "createdAt": 345600000,
            "description": "credit card",
            "id": "3",
            "note": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "expense list",
          },
          "ref": null,
          "rendered": "expense list",
          "type": "h1",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`should render no expense p element with no expenses 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ExpenseList
    expenses={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <p>
           No Expenses 
        </p>,
        <h1>
          expense list
        </h1>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": " No Expenses ",
        },
        "ref": null,
        "rendered": " No Expenses ",
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "expense list",
        },
        "ref": null,
        "rendered": "expense list",
        "type": "h1",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <p>
             No Expenses 
          </p>,
          <h1>
            expense list
          </h1>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": " No Expenses ",
          },
          "ref": null,
          "rendered": " No Expenses ",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "expense list",
          },
          "ref": null,
          "rendered": "expense list",
          "type": "h1",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
